documentation/cadl-structure-guidelines.md

```diff
@@ -0,0 +1,127 @@
+
+### Packages
+
+All services and service family libraries are modeled as Cadl packages, since you must install supporting librarie for proper tooling support. Per this proposal, all packages defined in the repo would be **unpublished**. Only packages in `cadl-azure` would be published. We will annotate the packages with `"private": true` in the `package.json` file to prevent publishing these to npm.
+
+All packages defined in this repo would use the `@cadl-api-spec` scope and use a lowercased, dashed form of the service namespace (ex: `@cadl-api-spec/azure-storage-blob`).
+
+### Structure
+
+Each package should have the following minimum structure:
+
+- `cadl-project.yaml`
+- `package.json`
+- `main.cadl` file
+- Supporting `*.cadl` files
+- `examples/` folder for example JSON files
+
+Authors may use folders as desired for organizing cadl files.
+
+Additionally, packages which wish to define custom linter rules or otherwise use TypeScript must place any `*.ts` files under a `src/` folder. Within the `src/` folder, the author may use as many subfolders as desired for organizing code.
+
+To distinguish between folders which define a service, an SDK, or both, one can look to the `cadl-project.yaml` and/or the `package.json`.
+
+- SDKs will take dependencies on the `@azure-tools/cadl-dpg` library, as well as SDK-specific emitters such as `@azure-tools/cadl-python` and configure them within `cadl-project.yaml`.
+- SDKs _may_ have a sidecar file to customize how the SDK is shaped. Folders that describe service definitions only will not have a sidecar file. _Note: the absence of a sidecar does not mean that a folder does not describe an SDK, but the presence of one means it is an SDK._
+- Services will take a dependency on `@azure-tools/autorest` and configure the emitter in `cadl-project.yaml`.
+
+## Service Family Libraries
+
+The service family library concept allows a family of services to share common models, linter rules, templates, etc.
+
+Service libraries can include unpublished service _family_ libraries via source dependency annotations in `package.json`:
+
+```json
+"peerDependencies": {
+  "@azure-tools/cadl-azure-core": "~x.x.x",
+  "@cadl-api-spec/azure-communication-common": "file: ../Communication.Shared"
+},
+```
+
+While this would permit services from importing any service library described in the specs repo, as a matter of policy we should probably avoid that and have tooling to detect this scenario. Service family libraries **should** use versioning decorators and spec packages should reference them as versioned dependencies. Tooling would need to ensure that changes to service family libraries does not result in unexpected changes to any service version. One way to do this would be to diff the projection of the service versions on the `main` branch against the projection of the service versions that result from the change.
+
```

Comment:
Do you have idea of which tool is supposed to calculate the diff between the PR changes and main branch?

Reply:
Eventually I'm hoping this will be the cadl tooling itself, probably the swagger emitter set of tooling.